Android/app/src/main/java/app/intra/sys/firebase/RemoteConfig.java

```diff
@@ -36,16 +38,50 @@
     }
   }
 
+  private static boolean goDohDisabled() {
+    try {
+      return FirebaseRemoteConfig.getInstance().getBoolean("disable_go_doh");
+    } catch (IllegalStateException e) {
+      LogWrapper.logException(e);
+      return false;
+    }
+  }
+
   public static boolean getUseGoDoh() {
-    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
-      // Must be kept in sync with the check in IntraVpnService.makeVpnAdapter.
-      // GoVpnAdapter isn't used on versions below M, so Go-DOH shouldn't be used for other purposes
-      // (especially probes) on those versions either.
+    if (getUseSplitMode()) {
+      // Split mode doesn't use Go-DOH, so Go-DOH shouldn't be used for other purposes
+      // (especially probes) in split mode, for consistency.
       return false;
     }
+    return !goDohDisabled();
+  }
+
+  public static boolean getUseSplitMode() {
+    // On M and later, Chrome uses getActiveNetwork() to determine which DNS servers to use.
+    // A full tunnel configuration makes this VPN the active network, whereas a
+    // split-tunnel configuration does not.  Therefore, on M and later, we cannot use
+    // split mode.  Additionally, M and later also exhibit DownloadManager bugs when used
+    // with a split-tunnel VPN.
+    if (VERSION.SDK_INT >= VERSION_CODES.M) {
+      return false;
+    }
+
+    if (VERSION.SDK_INT < VERSION_CODES.LOLLIPOP && goDohDisabled()) {
+      // Versions below Lollipop can only use full-VPN in combination with Go-DOH, because
+      // StandardServerConnection does not use socket protection.  (It also relies on
+      // InetAddress.getAllByName(), which cannot be protected.)
+      return true;
+    }
+
     try {
-      return !FirebaseRemoteConfig.getInstance()
-          .getBoolean("disable_go_doh");
+      // go_min_version is the minimum Android version that should use the GoVpnAdapter (i.e. the
+      // full tunnel configuration).
+      long minVersion = FirebaseRemoteConfig.getInstance().getLong("go_min_version");
```

Comment:
Should the min Go version be defined in source instead than remotely?

Reply:
The goal of remote configuration is to enable A/B testing and remote-config experiments, to test whether this codepath really works well in practice.  In principle (and in the available emulators), it works on all versions, but I have some concerns about memory usage and unexpected system behaviors.

Once we've established a min version (hopefully 15, which is Intra's min version overall), we can remove this code (and also the OkHttp DoH client).

Reply:
Got it!